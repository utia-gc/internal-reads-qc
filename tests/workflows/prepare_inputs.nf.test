nextflow_workflow {

    name "Test Workflow PREPARE_INPUTS"
    script "workflows/prepare_inputs.nf"
    workflow "PREPARE_INPUTS"

    test("PREPARE_INPUTS takes a directory to reads and emits read pairs with sample metadata -- SE + PE reads.") {

        when {
            params {
                // define parameters here. Example:
                // outdir = "tests/results"
            }
            workflow {
                """
                input[0] = "${projectDir}/tests/data/reads/raw"
                """
            }
        }

        then {
            // test workflow status
            assert workflow.success
            assert workflow.trace.succeeded().size() == 0

            // test reads parsed with metadata
            with(workflow.out.reads) {
                assert size() == 4

                // test SRR1066657 L001 SE reads
                with(get(0)) {
                    // test that format is [metadata, reads1, reads2]
                    assert size() == 3
                    assert get(0) instanceof Map
                    def reads1 = path(get(1)).toFile()
                    def reads2 = path(get(2)).toFile()
                    assert reads1.isFile()
                    assert reads2.isFile()

                    // test metadata fields
                    with(get(0)) {
                        assert sampleName   == 'SRR1066657'
                        assert readType     == 'single'
                        assert lane         == '001'
                    }
                    // test R1
                    assert reads1.toString() ==~ /^.*\/SRR1066657_S3_L001_R1_001\.fastq\.gz/
                    assert reads1.exists() && reads1.length() > 0
                    // test R2
                    assert reads2.toString() ==~ /^.*\/SRR1066657_S3_L001_R2_001\.fastq\.gz\.EMPTYFILE/
                    assert reads2.exists() && reads2.length() == 0
                }

                // test SRR1066657 L002 SE reads
                with(get(1)) {
                    // test that format is [metadata, reads1, reads2]
                    assert size() == 3
                    assert get(0) instanceof Map
                    def reads1 = path(get(1)).toFile()
                    def reads2 = path(get(2)).toFile()
                    assert reads1.isFile()
                    assert reads2.isFile()

                    // test metadata fields
                    with(get(0)) {
                        assert sampleName   == 'SRR1066657'
                        assert readType     == 'single'
                        assert lane         == '002'
                    }
                    // test R1
                    assert reads1.toString() ==~ /^.*\/SRR1066657_S3_L002_R1_001\.fastq\.gz/
                    assert reads1.exists() && reads1.length() > 0
                    // test R2
                    assert reads2.toString() ==~ /^.*\/SRR1066657_S3_L002_R2_001\.fastq\.gz\.EMPTYFILE/
                    assert reads2.exists() && reads2.length() == 0
                }

                // test SRR6924569 L001 PE reads
                with(get(2)) {
                    // test that format is [metadata, reads1, reads2]
                    assert size() == 3
                    assert get(0) instanceof Map
                    def reads1 = path(get(1)).toFile()
                    def reads2 = path(get(2)).toFile()
                    assert reads1.isFile()
                    assert reads2.isFile()

                    // test metadata fields
                    with(get(0)) {
                        assert sampleName   == 'SRR6924569'
                        assert readType     == 'paired'
                        assert lane         == '001'
                    }
                    // test R1
                    assert reads1.toString() ==~ /^.*\/SRR6924569_S1_L001_R1_001\.fastq\.gz/
                    assert reads1.exists() && reads1.length() > 0
                    // test R2
                    assert reads2 ==~ /^.*\/SRR6924569_S1_L001_R2_001\.fastq\.gz/
                    assert reads2.exists() && reads2.length() > 0
                }

                // test SRR6924569 L002 PE reads
                with(get(3)) {
                    // test that format is [metadata, reads1, reads2]
                    assert size() == 3
                    assert get(0) instanceof Map
                    def reads1 = path(get(1)).toFile()
                    def reads2 = path(get(2)).toFile()
                    assert reads1.isFile()
                    assert reads2.isFile()

                    // test metadata fields
                    with(get(0)) {
                        assert sampleName   == 'SRR6924569'
                        assert readType     == 'paired'
                        assert lane         == '002'
                    }
                    // test R1
                    assert reads1.toString() ==~ /^.*\/SRR6924569_S1_L002_R1_001\.fastq\.gz/
                    assert reads1.exists() && reads1.length() > 0
                    // test R2
                    assert reads2.toString() ==~ /^.*\/SRR6924569_S1_L002_R2_001\.fastq\.gz/
                    assert reads2.exists() && reads2.length() > 0
                }
            }

            // test workflow reproducibility
            assert snapshot(workflow.out).match()
        }

    }

}
