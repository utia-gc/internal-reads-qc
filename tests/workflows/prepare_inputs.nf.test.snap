{
    "PREPARE_INPUTS takes a directory to reads and emits read pairs with sample metadata -- SE + PE reads.": {
        "content": [
            {
                "0": [
                    [
                        {
                            "sampleName": "SRR0000000",
                            "lane": "",
                            "readType": "paired"
                        },
                        "SRR0000000_S8_R1_001.fastq.gz:md5,ca21995dbc0501708eba59cd04df1f81",
                        "SRR0000000_S8_R2_001.fastq.gz:md5,6748363463e372ce7cfc2d97633b492f"
                    ],
                    [
                        {
                            "sampleName": "SRR1066657",
                            "lane": "001",
                            "readType": "single"
                        },
                        "SRR1066657_S3_L001_R1_001.fastq.gz:md5,2a288d18d3b65cd2d33bb51a9b1c3aa5",
                        "SRR1066657_S3_L001_R2_001.fastq.gz.EMPTYFILE:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ],
                    [
                        {
                            "sampleName": "SRR1066657",
                            "lane": "002",
                            "readType": "single"
                        },
                        "SRR1066657_S3_L002_R1_001.fastq.gz:md5,24e26bcb7fe0ac19e41664382662d166",
                        "SRR1066657_S3_L002_R2_001.fastq.gz.EMPTYFILE:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ],
                    [
                        {
                            "sampleName": "SRR6924569",
                            "lane": "001",
                            "readType": "paired"
                        },
                        "SRR6924569_S1_L001_R1_001.fastq.gz:md5,ca21995dbc0501708eba59cd04df1f81",
                        "SRR6924569_S1_L001_R2_001.fastq.gz:md5,6748363463e372ce7cfc2d97633b492f"
                    ],
                    [
                        {
                            "sampleName": "SRR6924569",
                            "lane": "002",
                            "readType": "paired"
                        },
                        "SRR6924569_S1_L002_R1_001.fastq.gz:md5,bd0746d64b03d97b27c6a8df94565836",
                        "SRR6924569_S1_L002_R2_001.fastq.gz:md5,e77b0abbf4abdc0afcddf778f902db32"
                    ]
                ],
                "reads": [
                    [
                        {
                            "sampleName": "SRR0000000",
                            "lane": "",
                            "readType": "paired"
                        },
                        "SRR0000000_S8_R1_001.fastq.gz:md5,ca21995dbc0501708eba59cd04df1f81",
                        "SRR0000000_S8_R2_001.fastq.gz:md5,6748363463e372ce7cfc2d97633b492f"
                    ],
                    [
                        {
                            "sampleName": "SRR1066657",
                            "lane": "001",
                            "readType": "single"
                        },
                        "SRR1066657_S3_L001_R1_001.fastq.gz:md5,2a288d18d3b65cd2d33bb51a9b1c3aa5",
                        "SRR1066657_S3_L001_R2_001.fastq.gz.EMPTYFILE:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ],
                    [
                        {
                            "sampleName": "SRR1066657",
                            "lane": "002",
                            "readType": "single"
                        },
                        "SRR1066657_S3_L002_R1_001.fastq.gz:md5,24e26bcb7fe0ac19e41664382662d166",
                        "SRR1066657_S3_L002_R2_001.fastq.gz.EMPTYFILE:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ],
                    [
                        {
                            "sampleName": "SRR6924569",
                            "lane": "001",
                            "readType": "paired"
                        },
                        "SRR6924569_S1_L001_R1_001.fastq.gz:md5,ca21995dbc0501708eba59cd04df1f81",
                        "SRR6924569_S1_L001_R2_001.fastq.gz:md5,6748363463e372ce7cfc2d97633b492f"
                    ],
                    [
                        {
                            "sampleName": "SRR6924569",
                            "lane": "002",
                            "readType": "paired"
                        },
                        "SRR6924569_S1_L002_R1_001.fastq.gz:md5,bd0746d64b03d97b27c6a8df94565836",
                        "SRR6924569_S1_L002_R2_001.fastq.gz:md5,e77b0abbf4abdc0afcddf778f902db32"
                    ]
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2025-03-06T17:04:34.397928672"
    }
}